File: mobile/lib/models/member.dart
import 'package:equatable/equatable.dart';

class Member extends Equatable {
  final int id;
  final String memberCode;
  final String name;
  final String phone;
  final String email;
  final String address;
  final String idNumber;
  final String type; // regular, silver, gold, platinum
  final int points;
  final double totalPurchase;
  final double totalSell;
  final DateTime? joinDate;
  final DateTime? birthDate;
  final String notes;
  final bool isActive;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Member({
    required this.id,
    required this.memberCode,
    required this.name,
    this.phone = '',
    this.email = '',
    this.address = '',
    this.idNumber = '',
    this.type = 'regular',
    this.points = 0,
    this.totalPurchase = 0,
    this.totalSell = 0,
    this.joinDate,
    this.birthDate,
    this.notes = '',
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
  });

  // Alias for consistency
  int get totalPurchases => totalPurchase.toInt();

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'] ?? 0,
      memberCode: json['member_code'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      email: json['email'] ?? '',
      address: json['address'] ?? '',
      idNumber: json['id_number'] ?? '',
      type: json['type'] ?? 'regular',
      points: json['points'] ?? 0,
      totalPurchase: (json['total_purchase'] ?? 0).toDouble(),
      totalSell: (json['total_sell'] ?? 0).toDouble(),
      joinDate: json['join_date'] != null ? DateTime.parse(json['join_date']) : null,
      birthDate: json['birth_date'] != null ? DateTime.parse(json['birth_date']) : null,
      notes: json['notes'] ?? '',
      isActive: json['is_active'] ?? true,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'member_code': memberCode,
      'name': name,
      'phone': phone,
      'email': email,
      'address': address,
      'id_number': idNumber,
      'type': type,
      'points': points,
      'total_purchase': totalPurchase,
      'total_sell': totalSell,
      'join_date': joinDate?.toIso8601String(),
      'birth_date': birthDate?.toIso8601String(),
      'notes': notes,
      'is_active': isActive,
    };
  }

  Member copyWith({
    int? id,
    String? memberCode,
    String? name,
    String? phone,
    String? email,
    String? address,
    String? idNumber,
    String? type,
    int? points,
    double? totalPurchase,
    double? totalSell,
    DateTime? joinDate,
    DateTime? birthDate,
    String? notes,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Member(
      id: id ?? this.id,
      memberCode: memberCode ?? this.memberCode,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      email: email ?? this.email,
      address: address ?? this.address,
      idNumber: idNumber ?? this.idNumber,
      type: type ?? this.type,
      points: points ?? this.points,
      totalPurchase: totalPurchase ?? this.totalPurchase,
      totalSell: totalSell ?? this.totalSell,
      joinDate: joinDate ?? this.joinDate,
      birthDate: birthDate ?? this.birthDate,
      notes: notes ?? this.notes,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [id, memberCode, name, phone, type, isActive];
}

File: backend/models/member.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// MemberType defines the type of member
type MemberType string

const (
	MemberTypeRegular  MemberType = "regular"
	MemberTypeSilver   MemberType = "silver"
	MemberTypeGold     MemberType = "gold"
	MemberTypePlatinum MemberType = "platinum"
)

// Member represents customer/member data
type Member struct {
	ID               uint           `gorm:"primarykey" json:"id"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`
	MemberCode       string         `gorm:"not null;size:20" json:"member_code"` // unique index created manually in migration
	Name             string         `gorm:"not null;size:100" json:"name"`
	Phone            string         `gorm:"size:20;index" json:"phone"`
	Email            string         `gorm:"size:100" json:"email"`
	Address          string         `gorm:"size:255" json:"address"`
	IDNumber         string         `gorm:"size:30" json:"id_number"` // KTP/ID Card number
	Type             MemberType     `gorm:"not null;size:20;default:'regular'" json:"type"`
	Points           int            `gorm:"default:0" json:"points"`
	TotalPurchase    float64        `gorm:"default:0" json:"total_purchase"`    // Total beli dari toko (sale)
	TotalSell        float64        `gorm:"default:0" json:"total_sell"`        // Total jual ke toko (setor emas)
	TransactionCount int            `gorm:"default:0" json:"transaction_count"` // Total jumlah transaksi
	JoinDate         time.Time      `json:"join_date"`
	BirthDate        *time.Time     `json:"birth_date,omitempty"`
	Notes            string         `gorm:"size:500" json:"notes"`
	IsActive         bool           `gorm:"default:true" json:"is_active"`

	// Relations
	Transactions []Transaction `gorm:"foreignKey:MemberID" json:"transactions,omitempty"`
}

// AddPoints adds loyalty points to member
func (m *Member) AddPoints(amount float64) {
	// Example: 1 point per 100,000 spent
	points := int(amount / 100000)
	m.Points += points

	// Update member type based on total purchase
	m.updateMemberType()
}

// updateMemberType updates member type based on total purchase
func (m *Member) updateMemberType() {
	if m.TotalPurchase >= 100000000 { // 100 juta
		m.Type = MemberTypePlatinum
	} else if m.TotalPurchase >= 50000000 { // 50 juta
		m.Type = MemberTypeGold
	} else if m.TotalPurchase >= 20000000 { // 20 juta
		m.Type = MemberTypeSilver
	} else {
		m.Type = MemberTypeRegular
	}
}
